### PR TITLE
Configuration faster

### PR DESCRIPTION
**Pull request by jbogard**
Made configuration faster, about 2-3x in a local project. @TylerCarlson1 I poked a bit around the convention stuff to precalculate some stuff that was getting calculated every time.

### CODE REVIEW

**Comment by TylerCarlson1** [comment]
Yea looks good for the most part.  Have a hard time following Caching and Pre-Calculating config options but I assume copy paste refactor worked fine.

**Comment by jbogard** [comment]
I thought about making the PrePost fix class do all the caching, but it was expensive to do the lookups vs. just putting the results on the TypeDetails. Then I also saw that we create TypeDetails (which is expensive) based on a profile, so why not cache them in the profile.

**Comment by TylerCarlson1** [comment]
can you cache them in a dictionary that way the are cached for whatever part of Automapper uses them like you do with the expressions and typepair to maps?

**Comment by jbogard** [comment]
Eh?

**Comment by TylerCarlson1** [comment]
oh you are already doing it.  I saw CreateTypeDetails and forgot it cached.

File: src/AutoMapper/Mappers/DynamicMappers.cs [hunk 1]
```diff
@@ -63,7 +63,7 @@ public class ToDynamicMapper : IObjectMapper
         {
             if (destination == null)
                 destination = ifNull();
-            var sourceTypeDetails = new TypeDetails(typeof(TSource), context.ConfigurationProvider.Configuration);
+            var sourceTypeDetails = context.ConfigurationProvider.Configuration.CreateTypeDetails(typeof(TSource));
```
Comment (lbargaoanu) [inline src/AutoMapper/Mappers/DynamicMappers.cs]: Maybe expose configuration from the context directly.

**Pull request merged**